### PR TITLE
BTFS-2020 BTFS-2021 BTFS-2025

### PR DESCRIPTION
**Pull request by laipogo**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior?** (You can also refer to a JIRA ticket here)

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)

* **Description of changes**

---

* **Please check if the PR fulfills these requirements**
- [ ] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ ] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ ] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [ ] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [ ] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [ ] Code changes have run through `go fmt`
- [ ] Code changes have run through `go mod tidy`
- [ ] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

File: core/commands/wallet.go [hunk 1]
```diff
@@ -190,19 +195,19 @@ var walletPasswordCmd = &cmds.Command{
...
-		}
-		n, err := cmdenv.GetNode(env)
+		cfg, err := n.Repo.Config()
 		if err != nil {
 			return err
 		}
-		cfg.Identity.Password = req.Arguments[0]
-		cfg.UI.Wallet.Initialized = true
+		if cfg.UI.Wallet.Initialized {
+			return errors.New("Already init, cannot set pasword again.")
```
Comment (taiyangc) [inline core/commands/wallet.go:207]: password

File: protos/wallet/transaction.proto [hunk 2]
```diff
@@ -0,0 +1,33 @@
+syntax = "proto3";
```
Comment (taiyangc) [inline protos/wallet/transaction.proto:1]: remove since transactions not supported (yet)

File: core/commands/wallet.go [hunk 3]
```diff
@@ -91,6 +93,9 @@ var walletDepositCmd = &cmds.Command{
 		err = wallet.WalletDeposit(cfg, amount, runDaemon)
 		if err != nil {
 			log.Error("wallet deposit failed, ERR: ", err)
```
Comment (taiyangc) [inline core/commands/wallet.go:95]: do not double log - it will be confusing

File: core/commands/wallet.go [hunk 4]
```diff
@@ -132,6 +137,9 @@ var walletWithdrawCmd = &cmds.Command{
 		err = wallet.WalletWithdraw(cfg, amount)
 		if err != nil {
 			log.Error("wallet withdraw failed, ERR: ", err)
```
Comment (taiyangc) [inline core/commands/wallet.go:139]: do not double log - it will be confusing

File: core/commands/wallet.go [hunk 5]
```diff
@@ -63,6 +65,7 @@ var walletDepositCmd = &cmds.Command{
 	Helptext: cmds.HelpText{
 		Tagline:          "BTFS wallet deposit",
 		ShortDescription: "BTFS wallet deposit from block chain to ledger.",
+		Options:          "unit is µ (=0.0000001BTT)",
```
Comment (taiyangc) [inline core/commands/wallet.go:68]: 0.000001 BTT

Comment (taiyangc) [inline core/commands/wallet.go:68]: unit is `µBTT`

File: core/commands/wallet.go [hunk 6]
```diff
@@ -90,7 +93,9 @@ var walletDepositCmd = &cmds.Command{
 
 		err = wallet.WalletDeposit(cfg, amount, runDaemon)
 		if err != nil {
-			log.Error("wallet deposit failed, ERR: ", err)
+			if strings.Contains(err.Error(), "Please deposit at least") {
+				err = errors.New("Please deposit at least 10,000,000µ(=10BTT)")
```
Comment (taiyangc) [inline core/commands/wallet.go:97]: µBTT

File: core/commands/wallet.go [hunk 7]
```diff
@@ -151,33 +159,39 @@ var walletBalanceCmd = &cmds.Command{
...
-
-		return cmds.EmitOnce(res, &MessageOutput{s})
-	},
-	Encoders: cmds.EncoderMap{
-		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *MessageOutput) error {
-			fmt.Fprint(w, out.Message)
-			return nil
-		}),
+		return cmds.EmitOnce(res, &BalanceResponse{
+			BtfsWalletBalance: fmt.Sprintf("%dµ", tronBalance),
```
Comment (taiyangc) [inline core/commands/wallet.go:185]: balance should be int according to doc.

File: core/commands/wallet.go [hunk 8]
```diff
@@ -151,33 +159,39 @@ var walletBalanceCmd = &cmds.Command{
...
+			BttWalletBalance:  fmt.Sprintf("%dµ", ledgerBalance),
+		})
 	},
-	Type: MessageOutput{},
+	Type: BalanceResponse{},
+}
+
+type BalanceResponse struct {
+	BtfsWalletBalance string
+	BttWalletBalance  string
```
Comment (taiyangc) [inline core/commands/wallet.go:194]: uint64

File: core/wallet/aes.go [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+var iv = []byte{0x02, 0x00, 0x01, 0x06, 0x00, 0x08, 0x01, 0x04, 0x02, 0x00, 0x01, 0x06, 0x00, 0x08, 0x01, 0x04}
+
+func EncryptWithAES(key, message string) string {
+	hash := md5.New()
+	hash.Write([]byte(key))
+	keyData := hash.Sum(nil)
+
+	block, err := aes.NewCipher(keyData)
+	if err != nil {
+		panic(err)
```
Comment (taiyangc) [inline core/wallet/aes.go:20]: normal functions within commands should not panic. needs to display error

File: core/commands/wallet.go [hunk 10]
```diff
@@ -190,19 +204,27 @@ var walletPasswordCmd = &cmds.Command{
...
+		cfg, err := n.Repo.Config()
 		if err != nil {
 			return err
 		}
 		if cfg.UI.Wallet.Initialized {
-			return errors.New("Already init, cannot set pasword again.")
+			return errors.New("Already init, cannot set password again.")
 		}
-		n, err := cmdenv.GetNode(env)
+		cipherMnemonic, err := wallet.EncryptWithAES(req.Arguments[0], cfg.Identity.Mnemonic)
```
Comment (JustinGaoF) [inline core/commands/wallet.go:218]: not sure, seems req.Arguments[1] was not used here.

Comment (taiyangc) [reply core/commands/wallet.go:218]: There is only one argument.. req.Arguments[0] is the password

**Pull request merged**